src/accel/SharedParams.cc

```diff
@@ -29,64 +29,76 @@
 #include "celeritas/random/RngParams.hh"
 #include "celeritas/track/TrackInitParams.hh"
 
+#include "SetupOptions.hh"
+
 namespace celeritas
 {
 //---------------------------------------------------------------------------//
 /*!
- * Construct with Celeritas setup options and shared data.
+ * Construct a shared pointer to this class for sharing across threads.
  */
-RunAction::RunAction(SPConstOptions options,
-                     SPParams       params,
-                     SPTransporter  transport)
-    : options_(options), params_(params), transport_(transport)
+std::shared_ptr<SharedParams> SharedParams::MakeShared()
 {
-    CELER_EXPECT(options_);
-    CELER_EXPECT(params_);
-    CELER_EXPECT(transport_);
+    return std::make_shared<SharedParams>();
 }
 
+//---------------------------------------------------------------------------//
+//! Default destructor
+SharedParams::~SharedParams() = default;
+
 //---------------------------------------------------------------------------//
 /*!
- * Initialize Celeritas.
+ * Thread-safe setup of Celeritas using Geant4 data.
+ *
+ * This is a separate step from construction because it has to happen at the
+ * beginning of the run, not when user classes are created.
  */
-void RunAction::BeginOfRunAction(const G4Run* run)
+void SharedParams::Initialize(const SetupOptions& options)
 {
-    CELER_EXPECT(run);
+    if (Device::num_devices() > 0)
+    {
+        // Initialize CUDA (you'll need to use CUDA environment variables to
+        // control the preferred device)
+        celeritas::activate_device(Device{0});
```

Comment:
Is the device initialization thread-safe? (Looking at the linked discussions in Device.cc: https://github.com/celeritas-project/celeritas/pull/149#discussion_r577997723 and https://github.com/celeritas-project/celeritas/pull/149#discussion_r578000062)

Reply:
The `cudaSetDevice` call definitely needs to be called from each thread (see also https://github.com/celeritas-project/celeritas/issues/418#issuecomment-1303522256 ) but I think
```c++
    global_device() = std::move(device);
```
is not thread-safe. So maybe there needs to be a mutex there? Maybe `global_device()` needs to return a thread-local `Device`? Not sure...